Added settings for Meter Pro Co2

ahrmn opened the pull request:
I reverse-engineered the settings of the Meter Pro (CO2) and implemented them here.
I have no means of testing if this work for the other Meter models.
An example script to fully configure a device using this would be for example
```
import asyncio

from switchbot.devices.meter import SwitchbotMeter
from switchbot.discovery import GetSwitchbotDevices
from switchbot.const import SwitchbotModel

BLE_MAC='XX:XX:XX:XX:XX:XX'
METER_MODEL=SwitchbotModel.METER_PRO_C

async def main():
    switchbotDevices = await GetSwitchbotDevices().discover()
    meter_advertisement = switchbotDevices[BLE_MAC]
    print(meter_advertisement.data)
    meter = SwitchbotMeter(device=meter_advertisement.device, model=METER_MODEL)
    await meter.show_battery_level(False)
    await meter.set_co2_thresholds(900, 1200)
    await meter.set_comfortlevel(19, 22, 40, 70)
    await meter.set_temperature_update_interval(10)
    await meter.set_co2_update_interval(10)
    await meter.set_button_function(False, False)
    await meter.set_alert_sound(False, 2)
    await meter.set_alert_co2(True, 400, 1200, False)
    
asyncio.run(main())
```

Additionally one can trigger a new measurement regardless of the frequency using 
`await meter.force_new_CO2_measurement()` and calibrate the sensor (at fresh air!) using 
`await meter.calibrate_co2_sensor()`.

ahrmn committed:
Added settings for Meter Pro Co2

ahrmn commented:
This may solve [#416](https://github.com/sblibs/pySwitchbot/issues/416).
@jacekowski Can you test if this synchronises time for the Meter Pro (without CO2 sensor)?

zerzhang commented on switchbot/devices/meter.py:
```diff
@@ -1 +1,323 @@
+    async def show_battery_level(self, show_battery: bool):
+        """Show or hide battery level on the display."""
+        show_battery_byte = "01" if show_battery else "00"
+        await super()._send_command(COMMAND_SHOW_BATTERY_LEVEL + show_battery_byte)
```
00 indicates that the battery icon will be displayed when the battery level is <= 20%, 01 indicates the default display.

zerzhang commented on switchbot/devices/meter.py:
```diff
@@ -1 +1,323 @@
+        await super()._send_command(
+            COMMAND_COMFORTLEVEL
+            + hot_byte
+            + f"{wet:02x}"
```
bit7 is reserved,  bit6-0 represent the humidity.

zerzhang commented on switchbot/devices/meter.py:
```diff
@@ -1 +1,323 @@
+        point_five += 0x05
+    if int(hot * 10) % 10 == 5:
+        point_five += 0x50
+    return f"{point_five:02x}"
```
```
def get_temp_decimal_byte(cold: float, hot: float) -> int:
    cold_dec = int(round(cold * 10)) % 10
    hot_dec = int(round(hot * 10)) % 10

    if not (0 <= cold_dec <= 9 and 0 <= hot_dec <= 9):
        raise ValueError("Decimal part out of range")

    return (hot_dec << 4) | cold_dec
```

"Temperature ValueH decimal
bit [7-4]: Temperature ValueH decimal, upper limit of temperature alarm, decimal part
Temperature ValueL decimal
bit [3-0]: Temperature ValueL decimal, lower limit of temperature alarm, decimal part
0000 –1001: 0~9"

This section itself supports setting decimal numbers from 0 to 9.

zerzhang commented on switchbot/devices/meter.py:
```diff
@@ -1 +1,323 @@
+            + f"{wet:02x}"
+            + point_five
+            + cold_byte
+            + f"{dry:02x}"
```
same as wet byte

zerzhang commented on switchbot/devices/meter.py:
```diff
@@ -1 +1,323 @@
+        await super()._send_command(
+            COMMAND_ALERT_SOUND + f"{volume:02x}" + sound_on_byte
+        )
+
```
Alarm Volume
0xFF -hold
0x01 Off
0x02 Low
0x03 Medium (Default)
0x04 High

Audible and visual alarm mode
0x00 - disable
0x01 - Flashing only
0x02 - Sound + flashing

pre-commit-ci[bot] committed:
chore(pre-commit.ci): auto fixes

ahrmn committed:
Feedback and Tests

pre-commit-ci[bot] committed:
chore(pre-commit.ci): auto fixes

ahrmn committed:
Merge branch 'master' into master

ahrmn committed:
Removed Humidity and Temperature Alert

ahrmn committed:
Covererage

ahrmn committed:
Match Errormessage

pre-commit-ci[bot] committed:
chore(pre-commit.ci): auto fixes

ahrmn committed:
renamed tests

pre-commit-ci[bot] committed:
chore(pre-commit.ci): auto fixes

ahrmn committed:
escape

pre-commit-ci[bot] committed:
chore(pre-commit.ci): auto fixes

ahrmn commented:
@zerzhang May I ask what is blocking this?
I removed the Humidity- and Temperature alerts that caused problems with the CI and added more test-coverage.

zerzhang commented on switchbot/devices/meter_pro.py:
```diff
@@ -157,6 +179,127 @@ async def set_time_display_format(self, is_12h_mode: bool = False) -> None:
         result = await self._send_command(payload)
         self._validate_result("set_time_display_format", result)
 
+    async def show_battery_level(self, show_battery: bool):
```
Missing return type

zerzhang commented on switchbot/devices/meter_pro.py:
```diff
@@ -157,6 +179,127 @@ async def set_time_display_format(self, is_12h_mode: bool = False) -> None:
+        show_battery_byte = "01" if show_battery else "00"
+        await self._send_command(COMMAND_SHOW_BATTERY_LEVEL + show_battery_byte)
+
+    async def set_co2_thresholds(self, lower: int, upper: int):
```
Missing return type

zerzhang commented on switchbot/devices/meter_pro.py:
```diff
@@ -157,6 +179,127 @@ async def set_time_display_format(self, is_12h_mode: bool = False) -> None:
+            COMMAND_CO2_THRESHOLDS + f"{lower:04x}" + f"{upper:04x}"
+        )
+
+    async def set_comfortlevel(self, cold: float, hot: float, dry: int, wet: int):
```
Missing return type

zerzhang commented on switchbot/devices/meter_pro.py:
```diff
@@ -157,6 +179,235 @@ async def set_time_display_format(self, is_12h_mode: bool = False) -> None:
+        """
+        Sets the CO2-Alert.
+        on: Turn CO2-Alert on or off
+        lower and upper: The provided range (between 400ppm and 2000ppm in 100ppm steps)
```
The app has implemented restrictions; in reality, both the upper and lower limits are two bytes.

zerzhang commented on switchbot/devices/meter_pro.py:
```diff
@@ -157,6 +179,235 @@ async def set_time_display_format(self, is_12h_mode: bool = False) -> None:
+        Sets the interval in which temperature and humidity are measured in battery powered mode.
+        Original App assumes minutes in {5, 10, 30}
+        """
+        seconds = minutes * 60
```
Agreed, it is recommended to use seconds as the parameter.

zerzhang commented on switchbot/devices/meter_pro.py:
```diff
@@ -157,6 +179,127 @@ async def set_time_display_format(self, is_12h_mode: bool = False) -> None:
+            + f"{dry:02x}"
+        )
+
+    async def set_alert_co2(self, on: bool, co2_low: int, co2_high: int, reverse: bool):
```
Missing return type

zerzhang commented on switchbot/devices/meter_pro.py:
```diff
@@ -157,6 +179,127 @@ async def set_time_display_format(self, is_12h_mode: bool = False) -> None:
+            COMMAND_ALERT_CO2 + f"{mode:02x}" + f"{co2_high:04x}" + f"{co2_low:04x}"
+        )
+
+    async def set_temperature_update_interval(self, minutes: int):
```
Missing return type

zerzhang commented on switchbot/devices/meter_pro.py:
```diff
@@ -157,6 +179,127 @@ async def set_time_display_format(self, is_12h_mode: bool = False) -> None:
+        seconds = minutes * 60
+        await self._send_command(COMMAND_TEMPERATURE_UPDATE_INTERVAL + f"{seconds:04x}")
+
+    async def set_co2_update_interval(self, minutes: int):
```
Missing return type

zerzhang commented on switchbot/devices/meter_pro.py:
```diff
@@ -157,6 +179,127 @@ async def set_time_display_format(self, is_12h_mode: bool = False) -> None:
+        seconds = minutes * 60
+        await self._send_command(COMMAND_CO2_UPDATE_INTERVAL + f"{seconds:04x}")
+
+    async def set_button_function(self, change_unit: bool, change_data_source: bool):
```
Missing return type

zerzhang commented on switchbot/devices/meter_pro.py:
```diff
@@ -157,6 +179,127 @@ async def set_time_display_format(self, is_12h_mode: bool = False) -> None:
+            COMMAND_BUTTON_FUNCTION + change_unit_byte + change_data_source_byte
+        )
+
+    async def force_new_co2_measurement(self):
```
Missing return type

zerzhang commented on switchbot/devices/meter_pro.py:
```diff
@@ -157,6 +179,127 @@ async def set_time_display_format(self, is_12h_mode: bool = False) -> None:
+        """Requests a new CO2 measurement, regardless of update interval"""
+        await self._send_command(COMMAND_FORCE_NEW_CO2_MEASUREMENT)
+
+    async def calibrate_co2_sensor(self):
```
Missing return type

zerzhang commented on switchbot/devices/meter_pro.py:
```diff
@@ -157,6 +179,127 @@ async def set_time_display_format(self, is_12h_mode: bool = False) -> None:
+        """
+        await self._send_command(COMMAND_CALIBRATE_CO2_SENSOR)
+
+    async def set_alert_sound(self, sound_on: bool, volume: int):
```
Missing return type

bdraco committed:
Merge branch 'master' into master

ahrmn commented:
@zerzhang Thanks for the Feedback. It took a while to find time for it.
I processed parts of it. But some comments i don't understand.
And I don't know what to do with the excel.
And yes, I very much regret not using a Feature-Branch :sweat_smile: 

Since this this does not look like it could be merged anytime soon, i decided to open a much smaller and more manageable PR #559
